doc: clarify the documentation of `Assume` assertion (#32100)
bitcoin/bitcoin: doc/developer-notes.md

  safely continue even if the assumption is violated. In debug builds it
  behaves like `Assert`/`assert` to notify developers and testers about
  nonfatal errors. In production it doesn't warn or log anything, though the

>   expression is always evaluated. However, if the compiler can prove that a
>   statement inside `Assume` is side-effect-free, it may optimize the call away,

Reviewer: According to https://www.reddit.com/r/cpp_questions/comments/plg8ij/expression_vs_statement/
> Statements do not have a return type

So by definition, unless they're empty, they cannot be `side-effect-free` - if my interpretation is correct.
```suggestion
  expression is always evaluated. However, if the compiler can prove that
  an expression inside `Assume` is side-effect-free, it may optimize the call away,
```

Author: Fixed!

>   an expression inside `Assume` is side-effect-free, it may optimize the call away,

Reviewer 2: ```suggestion
  an expression inside `Assume` is true, it _will_ have no affect,
```

Reviewer 2: s/affect/effect

Author: will fix not when there is need to retouch.

Reviewer 2: > If the expression inside Assume has a side-effect, but still can be proven to evaluate to true, the side effect will still remain.

I am assuming you are talking about the Macro then and not the C++ 23 [Assume](http://fiona.dmcs.p.lodz.pl/oopc/reference/en/cpp/language/attributes/assume.html) since that's not how the docs read.  The docs read that if it's true it has no effect: "If the expression would not evaluate to true at the place the assumption occurs, the behavior is undefined. Otherwise, the statement has no effect. "

Also, I think _may_ should be replaced with _will_ unless it's ambigous wether or not this will be otimized away depending on it's side-effect which it doesn't sound like it is.

Reviewer 2: I just noticed that Sipa said "There is no guarantee that the compiler will do that." which I guess is why the _may_ exists here.

Reviewer 2: > Yes, nothing in this PR is about the C++23 [[assume()]] attribute. I don't understand why you keep bringing it up.

The name collision had me confused

  skipping its evaluation in production. This enables a lower-cost way of

>   making explicit statements about the code, aiding review.

Reviewer 3: Perhaps keep the "In production" at the beginning of the sentence.

```
  In production builds, however, if the compiler can prove that
  an expression inside `Assume` is side-effect-free, it may optimize the call away,
  skipping its evaluation. 
```

  safely continue even if the assumption is violated. In debug builds it
  behaves like `Assert`/`assert` to notify developers and testers about
  nonfatal errors. In production it doesn't warn or log anything, though the
  expression is always evaluated. However, if the compiler can prove that a

>   statement inside `Assume` is side-effect-free, it may optimize the call away,

Reviewer 2: Interesting.  If `Assume` can be optimized out during production possibly, is there any reason to prefer `Assert`?

There's an assert [here](https://github.com/bitcoin/bitcoin/blob/998386d4462f5e06412303ba559791da83b913fb/src/wallet/coinselection.cpp#L159) in a coin-selection algo that is meant to be performant.  Maybe this should be changed to `Asume`.

Reviewer 2: I tested however the benchmarks don't seem to be sensitive enough for this to make a difference.

  skipping its evaluation in production.
   - `Assume` can also act as a lightweight debugging assertion, ensuring
     statements are tested e.g., during fuzzing—to catch violations. With
     sufficient test coverage, the compiler may optimize away proven conditions
     in production, eliminating runtime overhead. This aids both testing and code

>      review by making statements explicit and tested.

Author: fixed!

>      review by making statements explicit and tested.

Reviewer 4: Thank you for taking my suggestion!
